File: development/NET/samples/server/Session/CS/MySession.cs
using System;
using Softing.OPCToolbox;
using Softing.OPCToolbox.Server;

namespace Session
{
	class MySession : DaSession
	{

		public static MySession s_controlingSession = null;
		private static Int32 s_clientCount = 0;
		private ValueQT m_clientSpecValue = null;

		public MySession(EnumSessionType aType, uint aHandle) : base(aType, aHandle)
		{
			m_clientSpecValue = new ValueQT(s_clientCount, EnumQuality.GOOD, DateTime.Now);
			ClientNameChanged += new SessionClientNameChanged(HandleClientNameChanged);
			s_clientCount++;
		}   //	end ctr

		public uint Handle
		{
			get
			{
				return base.ObjectHandle;
			}   //	end get
		}   //	end Handle

		public override void ConnectSession()
		{

			base.ConnectSession();

			lock (Console.m_sessionsJanitor)
			{
				Console.m_sessions.Add(this);
			}
			Console.m_clientChanged.Set();

		}   //	end ConnectSession

		public override void DisconnectSession()
		{

			base.DisconnectSession();

			lock (Console.m_sessionsJanitor)
			{
				Console.m_sessions.Remove(this);
			}
			Console.m_clientChanged.Set();

		}   //	end DisconnectSession

		private void HandleClientNameChanged()
		{
			System.Console.WriteLine("New client name set for client {0:X}: {1}", Handle, ClientName);
		}   //	end HandleClientNameChanged

		public override int Logoff()
		{

			if (this.Equals(s_controlingSession))
			{
				s_controlingSession = null;
			}   //	end if

			Console.m_clientChanged.Set();
			return (int)EnumResultCode.S_OK;

		}   //	end Logoff

		public override int Logon(System.String UserName, System.String Password)
		{

			int ret = (int)EnumResultCode.E_ACCESSDENIED;

			if (UserName == "OPC" && Password == "opc")
			{

				if (this.Equals(s_controlingSession) || s_controlingSession == null)
				{
					s_controlingSession = this;
					ret = (int)EnumResultCode.S_OK;
				}   //	end if
			}   //	end if

			Console.m_clientChanged.Set();
			return ret;
		}   //	end Logon

		public int GetCacheValue(ref ValueQT aCacheValue)
		{

			aCacheValue = new ValueQT(m_clientSpecValue.Data, m_clientSpecValue.Quality, m_clientSpecValue.TimeStamp);
			return (int)EnumResultCode.S_OK;
		}   //	end GetCacheValue

	}   //	end class MySession
}

File: development/NET/samples/server/Session/VB/MySession.vb
Imports System
Imports Softing.OPCToolbox
Imports Softing.OPCToolbox.Server

Namespace Session
	Class MySession
		Inherits DaSession

#Region "Fields"
		Dim m_qualityGood As EnumQuality
#End Region

		Public Shared s_controlingSession As MySession = Nothing
		Private Shared s_clientCount As Int32 = 0
		Private m_clientSpecValue As ValueQT = Nothing

		Public Sub New(ByVal aType As EnumSessionType, ByVal aHandle As UInt32)
			MyBase.New(aType, aHandle)
			m_qualityGood = [Enum].ToObject(GetType(EnumQuality), EnumQuality.GOOD)
			m_clientSpecValue = New ValueQT(s_clientCount, m_qualityGood, DateTime.Now)
			System.Math.Max(System.Threading.Interlocked.Increment(s_clientCount), s_clientCount - 1)
			AddHandler ClientNameChanged, AddressOf HandleClientNameChanged
		End Sub

		Public ReadOnly Property Handle() As UInt32
			Get
				Return MyBase.ObjectHandle
			End Get
		End Property

		Public Overloads Overrides Sub ConnectSession()

			MyBase.ConnectSession()

			SyncLock Console.m_sessionsJanitor
				Console.m_sessions.Add(Me)
			End SyncLock

			Console.m_clientChanged.[Set]()

		End Sub

		Public Overloads Overrides Sub DisconnectSession()

			MyBase.DisconnectSession()

			SyncLock Console.m_sessionsJanitor
				Console.m_sessions.Remove(Me)
			End SyncLock

			Console.m_clientChanged.[Set]()

		End Sub

		Private Sub HandleClientNameChanged()
			System.Console.WriteLine("New client name set for client {0:X}: {1}", Handle, ClientName)
		End Sub

		Public Overloads Overrides Function Logoff() As Integer

			If Me.Equals(s_controlingSession) Then
				s_controlingSession = Nothing
			End If

			Console.m_clientChanged.[Set]()
			Return EnumResultCode.S_OK
		End Function

		Public Overloads Overrides Function Logon(ByVal UserName As String, ByVal Password As String) As Integer

			Dim ret As Integer = EnumResultCode.E_ACCESSDENIED

			If UserName = "OPC" AndAlso Password = "opc" Then

				If Me.Equals(s_controlingSession) OrElse s_controlingSession Is Nothing Then
					s_controlingSession = Me
					ret = EnumResultCode.S_OK
				End If
			End If

			Console.m_clientChanged.[Set]()
			Return ret
		End Function

		Public Function GetCacheValue(ByRef aCacheValue As ValueQT) As Integer

			aCacheValue = New ValueQT(m_clientSpecValue.Data, m_clientSpecValue.Quality, m_clientSpecValue.TimeStamp)
			Return EnumResultCode.S_OK
		End Function

	End Class
End Namespace

